segmentstore/server/src/main/java/io/pravega/segmentstore/server/writer/AttributeAggregator.java

```diff
@@ -0,0 +1,477 @@
+                        return this.dataSource.notifyAttributesPersisted(this.metadata.getId(), rpi.getRootPointer(), rpi.getLastSequenceNumber(), this.config.getFlushTimeout())
+                                .whenCompleteAsync((r, ex) -> {
+                                    if (ex != null) {
+                                        log.error("{}: Unable to persist root pointer {}.", this.traceObjectId, rpi, ex);
+                                    } else {
+                                        this.state.setLastPersistedSequenceNumber(rpi.getLastSequenceNumber());
+                                    }
+
+                                    // Set the latest value to null ONLY if it hasn't changed in the meantime.
+                                    if (this.lastRootPointer.compareAndSet(rpi, null)) {
+                                        // No new value. Instruct the loop to stop processing.
+                                        canContinue.set(false);
+                                    }
+                                }, this.executor);
+
+                    },
+                    this.executor);
+        }
+    }
+
+    /**
+     * Handles expected Attribute-related exceptions. Since the attribute index is a separate segment from the main one,
+     * it is highly likely that it may get temporarily out of sync with the main one, thus causing spurious StreamSegmentSealedExceptions
+     * or StreamSegmentNotExistsExceptions. If we get either of those, and they are consistent with our current state, the
+     * we can safely ignore them; otherwise we should be rethrowing them.
+     */
+    private <T> CompletableFuture<T> handleAttributeException(CompletableFuture<T> future) {
+        return Futures.exceptionallyExpecting(
+                future,
+                ex -> (ex instanceof StreamSegmentSealedException && this.metadata.isSealed())
+                        || ((ex instanceof StreamSegmentNotExistsException || ex instanceof StreamSegmentMergedException)
+                        && (this.metadata.isMerged() || this.metadata.isDeleted())),
+                null);
+    }
+
+    private boolean isSegmentDeleted() {
+        return this.metadata.isDeleted() || this.metadata.isMerged();
+    }
+
+    private Duration getElapsedSinceLastFlush() {
+        return this.timer.getElapsed().minus(this.lastFlush.get());
+    }
+
+    //endregion
+
+    //region RootPointer
+
+    @Data
+    private static class RootPointerInfo {
+        private final long rootPointer;
+        private final long lastSequenceNumber;
+
+        @Override
+        public String toString() {
+            return String.format("RootPointer=%s, LastSeqNo=%s", this.rootPointer, this.lastSequenceNumber);
+        }
+    }
+
+    //endregion
+
+    //region AggregatedAttributes
+
+    /**
+     * Aggregates pending Attribute Updates.
+     */
+    private static class State {
+        private final HashMap<UUID, Long> attributes;
```

Comment:
The other members of this class are atomics and this is a normal hashmap which is accessed by `include` without any synchronization. This looks like a bug.

Reply:
Good catch. Fixed.